File: backend/app/schemas/tenant.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class TenantItem(BaseModel):
    id: int
    code: str
    name: str
    is_active: bool
    session_version: int
    created_at: datetime
    updated_at: datetime


class TenantCreateRequest(BaseModel):
    code: str = Field(min_length=2, max_length=64)
    name: str = Field(min_length=1, max_length=120)


class TenantUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    is_active: bool | None = None


class TenantListData(BaseModel):
    items: list[TenantItem]
    total: int


class ModuleDefinitionItem(BaseModel):
    module_id: str
    name: str
    description: str | None = None
    icon: str | None = None
    menu_keys: list[str] = []
    permission_points: list[str] = []
    dependencies: list[str] = []
    is_default: bool = False
    sort_order: int = 0


class ModuleDefinitionListData(BaseModel):
    items: list[ModuleDefinitionItem]
    total: int


class TenantModuleItem(BaseModel):
    tenant_id: int
    module_id: str
    name: str
    description: str | None = None
    icon: str | None = None
    menu_keys: list[str] = []
    dependencies: list[str] = []
    depended_by: list[str] = []
    is_default: bool = False
    is_enabled: bool
    source: str
    enabled_at: datetime | None = None
    disabled_at: datetime | None = None
    sort_order: int = 0


class TenantModuleListData(BaseModel):
    tenant_id: int
    tenant_name: str
    items: list[TenantModuleItem]
    total: int


class TenantModuleUpdateRequest(BaseModel):
    module_id: str = Field(min_length=2, max_length=50)
    is_enabled: bool
    config_json: str | None = None


class ApiResponse(BaseModel):
    code: int
    message: str
    data: (
        TenantItem
        | TenantListData
        | ModuleDefinitionListData
        | TenantModuleListData
        | None
    ) = None

File: frontend/src/api/tenant.ts
import request from './request'

interface ApiResponse<T> {
  code: number
  message: string
  data: T
}

export interface TenantItem {
  id: number
  code: string
  name: string
  is_active: boolean
  session_version: number
  created_at: string
  updated_at: string
}

export interface ModuleDefinitionItem {
  module_id: string
  name: string
  description: string | null
  icon: string | null
  menu_keys: string[]
  permission_points: string[]
  dependencies: string[]
  is_default: boolean
  sort_order: number
}

export interface TenantModuleItem {
  tenant_id: number
  module_id: string
  name: string
  description: string | null
  icon: string | null
  menu_keys: string[]
  dependencies: string[]
  depended_by: string[]
  is_default: boolean
  is_enabled: boolean
  source: 'default' | 'tenant_override'
  enabled_at: string | null
  disabled_at: string | null
  sort_order: number
}

export interface TenantModuleListData {
  tenant_id: number
  tenant_name: string
  items: TenantModuleItem[]
  total: number
}

export const getTenants = async () => {
  const res = await request.get<ApiResponse<{ items: TenantItem[]; total: number }>>('/tenants')
  return res.data.data
}

export const createTenant = async (payload: { code: string; name: string }) => {
  const res = await request.post<ApiResponse<TenantItem>>('/tenants', payload)
  return res.data.data
}

export const updateTenant = async (tenantId: number, payload: { name?: string; is_active?: boolean }) => {
  const res = await request.put<ApiResponse<TenantItem>>(`/tenants/${tenantId}`, payload)
  return res.data.data
}

export const getModuleDefinitions = async () => {
  const res = await request.get<ApiResponse<{ items: ModuleDefinitionItem[]; total: number }>>('/tenants/modules')
  return res.data.data
}

export const getTenantModules = async (tenantId: number) => {
  const res = await request.get<ApiResponse<TenantModuleListData>>(`/tenants/${tenantId}/modules`)
  return res.data.data
}

export const updateTenantModule = async (
  tenantId: number,
  payload: { module_id: string; is_enabled: boolean; config_json?: string | null },
) => {
  const res = await request.put<ApiResponse<TenantModuleListData>>(`/tenants/${tenantId}/modules`, payload)
  return res.data.data
}
